### DISCUSSION TITLE
The Initial Learning Process in ASReview

### DISCUSSION THREAD

**Discussion by zhaokoh** [asreview/asreview#1200] [Q&A] [answered]
Hi ASReview team,

Thank you so much for creating this amazing tool, and keep improving it. v1.0 looks great and I am using it for my systematic review.

I am uploading a large number of data (probably 70% of them are already rated by human). Some details I would like to confirm with ASReview is: when the project is first initialised with this data (70% labeled, and 30% not labeled), does ASReview learns from all the labeled data, and then, re-ordered the 30% based on their relevance informed by the model?

The reason I asked is the initialisation process seems very efficient and fast with the following parameters in processing over approximately 65,000 titles and abstracts: 

- Feature extraction technique: TF-IDF
- Classifier: Logistic regression
- Query strategy: Maximum
- Balance strategy: Dynamic resampling

Thanks in advance.

**Comment by jteijema** [comment, marked as answer]
Hi @zhaokoh 

Good to hear you are enjoying the new v1.0 version! 

As for your questions; yes, when uploading a dataset with prelabeled data ASReview takes all labeled data as prior knowledge and learns from this for the next iterations.

As for the training speed, [here](https://psyarxiv.com/t7bpd) is a preprint paper containing a table with model processing times, based on a dataset with 46000 records. Logistic regression + TF-IDF is one of the fastest and best performing models available, completing initialization in under 24 seconds, and with a median iteration time of only 0.05 seconds.

Very efficient and fast is just right for these parameters!

Jelle

**Comment by zhaokoh** [comment]
Thanks Jelle for the information 🙏! I will keep an eye on the preprint and cite accordingly whenever my review is ready :).

**Comment by zhaokoh** [comment]
Hi ASReview team,

I would like to ask a further question in regard to this thread. Now I have around 90,000 including the original 65,000 articles above. So there are around 25000 articles that are undecided. I created a CSV file by:

1. Setting the 'Included' column to 0 for ineligible articles (among the 65,000)
2. Setting the 'Included' column to 1 for eligible articles (among the 65,000)
3. Then leaving 'Included' blank for the 25,000 articles.

To avoid order effects, I shuffled every row so the three different possibilities of 'Included' values are interspersed.

Next, I loaded this CSV file to ASReview (as a new project), and used the same learning configuration above. As usual, the learning process is super quick. Amazing! Let's call this Test 1.

Then, I performed the same steps (loaded CSV file to another new project with the same learning configuration). Let's call this Test 2.

I exported both datasets from Test 1 and Test 2.

I noticed that the exported datasets contain all the input rows, with the rankings added for both datasets. Some observations in the exported CSVs are:

- The input eligible rows (item 2 above) are placed at the top (ranked first)
- The ineligible rows (item 1 above) are placed at the bottom (ranked last).
- Then, I have the rows (without Included values) somewhere in between the two subsets above.

Now, the purpose of this exercise is to check how ASReview ranks those articles that do not contain a manual decision (Included is blank). I checked the ranking of those articles in the two sets. The rankings of the same row id between those two sets are different. For example, article A might have a ranking of 602 in Test 1 but 88 in Test 2. I am wondering if this is due to some sort of randomisation involved in the initialisation process?

(I am currently looking at the diff between the two rankings and see if they are minimal?).

Can you please let me know if this is expected and whether my observations are on the right track?

**Reply by jteijema** [reply]
Hey @zhaokoh 

To answer your question first; yes! This is expected behaviour. When using ASReview lab, no `seed` is set. While ASReview lab does have the `--seed` option, I do have to mention that this is part of Legacy code. If you want deterministic behaviour, then simulation mode is a saver bet.

Now, why do you see this difference? It is likely because the `Dynamic resampling balance strategy` samples random records, changing the behaviour of the model, and influencing what we call "rank order stability". The balancer samples records and creates a new dataset from those samples for the model to train on. In the two examples you tried, it is likely that the model trained on different datasets created from the same information, just sampled differently.

Many elements can influence the rank order stability, such as, but not limited to; the amount of prior knowledge, selection of models, quality and accuracy of labeling, quality of the information in the prior knowledge, how similar the subsets are in information, how much information the subset records contain, etc. 

The models in ASReview are designed to minimize the number of records that have to be screened to find all relevant records, and they work very well for this purpose. The models are not optimized for stabilizing the ranked order, as we are less interested in this. We just want to optimize the discovery rate, and actively choose this approach. Still, like you said, the difference is minimal when considering that it's position 88 and 602 out of 25000.

I hope this answers your question!

Jelle

**Comment by zhaokoh** [comment]
Thank you @JTeijema for this detailed explanation, very helpful! It sounds like `Dynamic resampling balance strategy` is purposely designed to target certain limitations with serial scanning articles (correct me if I am wrong here). In my scenario, I have a large set of prior knowledge, and I am interested in how the model 'predict' the unrated articles, in the form of ranking, based on the prior knowledge. I know that this is not a typical use case of ASReview but I am interested in trying using the tool this way. This morning I tried the 'Simple sampling strategy' and it seems to do what I wanted. Creating two projects with the same datasets (with the same prior knowledge) provides the same ranking for the same articles in both datasets.

If you don't mind me asking, pardon for my ignorance, I started reading some literature about semi-automated screening tools, and it seems some of these tools potentially have a shortcoming of 'hasty generalisation' problem. I am wondering if the dynamic sampling above is a way to reduce the likelihood of this issue?

Thanks again for your comment!

**Reply by Rensvandeschoot** [reply]
Super nice the `simple sampling` strategy behaved like you anticipated! 
The `dynamic resamplin` strategy was developed by @qubixes and @GerbrichFerdinands. It is part of the pre-print https://osf.io/w6qbg/ and the exact definition is available at https://osf.io/89wgm.

**Reply by zhaokoh** [reply]
Great @Rensvandeschoot for the response! I will have a look at the pre-print. Thanks again for making ASReview so easy to install, use and configure for a review project.
